### DISCUSSION TITLE
[General] Paymaster postTransaction did not get run if reverted

### DISCUSSION THREAD

**Discussion by nhynhyiddy** [zkSync-Community-Hub/zksync-developers#40] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

I am trying to write a Paymaster.sol which can do what paymaster does. I did implement some token transfer back to user to simulate refund. 

### What is the specific issue or error you're encountering?

I find out that if any code inside the function postTransaction is reverted, the transaction itself will not get reverted. It will just not run the entire postTransaction. Is it intended?

### Can you share the error messages or logs you're receiving, if any?

no error message. But i was expected the whole transaction to revert if postTransaction has a revert.

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

```
function postTransaction(
        bytes calldata _context,
        Transaction calldata _transaction,
        bytes32,
        bytes32,
        ExecutionResult _txResult,
        uint256 _maxRefundedGas
    ) external payable override onlyBootloader {
       revert("test revert");
    }
```

### Have you tried to isolate the problem, and if so, what were the results?

no

### What steps have you already taken to try to resolve the issue?

I just try to revert the postTransaction() but the transaction still runs ok. how do I force it to fail if there is an error?

**Comment by dutterbutter** [comment, marked as answer]
Hello 👋 

If you want to validate the transaction and potentially prevent it from being processed, you should do so in the `validateAndPayForPaymasterTransaction()` function. If this function reverts or fails, the transaction will not be processed.

To emphasize, the `validateAndPayForPaymasterTransaction()` function is the place to perform checks and validations before the transaction is executed, while the `postTransaction()` function is is called _after_ the transaction executes so the behaviour described is expected.

For a more detailed understanding of the transaction flow and the roles of these functions, I recommend reviewing the [documentation on our IPaymaster interface](https://era.zksync.io/docs/reference/concepts/aa.html#ipaymaster-interface).

**Reply by nhynhyiddy** [reply]
On what circumstances that the `postTransaction()` will not be called? I know out of gas is one. what about an error occurs within that function? 
In a blockchain transaction, normally any error within the transaction will fail the entire transaction, how did zkSync handle post transaction so that even if `postTransaction()` failed, the transaction will still run?

**Reply by dutterbutter** [reply]
Again, `postTransaction()` is called after the transaction executes, recall from the docs linked "...postTransaction takes four parameters: the context returned by validateAndPayForPaymasterTransaction, the transaction itself, a flag that indicates whether the transaction execution succeeded..." it is passed a flag whether the tx executed successfully. Consider it separated from transaction execution. 

Further, `postTransaction` will be called after a transaction unless there is not enough gas left to do so. If there's enough gas, `postTransaction` will be called. If any code inside `postTransaction` reverts it will just revert any effects which this method had. The transaction itself, however, will still be considered successful.

**Comment by steyn52** [comment]
In the context of a paymaster contract, if the postTransaction function reverts, it won't cause the entire transaction to revert. Instead, it will halt the execution of the paymaster logic without reverting the entire transaction. To force the entire transaction to revert if there is an error in postTransaction, you would need to implement additional checks or logic in your paymaster contract to handle this scenario. This could involve using require statements or other conditional checks to ensure that the transaction reverts if an error occurs in postTransaction.

**Comment by evmeiwear** [comment]
It seems like the behavior you're observing is expected. The postTransaction function in a paymaster contract acts as a callback and doesn't directly affect the execution of the transaction. If you want to revert the entire transaction based on conditions within postTransaction, you may need to handle that logic within the transaction itself rather than relying on the paymaster.

**Comment by mcleodjv** [comment]
In the postTransaction function of your Paymaster contract, the behavior you're experiencing where the transaction doesn't revert if there's a revert inside the function is likely due to the fact that the postTransaction function is called asynchronously after the transaction has already been executed. In this scenario, reverting inside postTransaction won't revert the entire transaction because the transaction execution has already completed. This behavior is intended and aligns with how the Paymaster pattern works in most cases. If you need the entire transaction to revert based on conditions inside postTransaction, you may need to reconsider your design or handle error conditions differently.

**Comment by casesumteryan** [comment]
In the Paymaster contract, the postTransaction function is intended to handle post-execution logic and is not meant to revert the entire transaction if an error occurs within it. If you want to revert the entire transaction upon encountering an error in postTransaction, you may need to implement additional logic in your contract to handle this scenario. One approach could be to emit an event or set a flag indicating an error occurred, and then check this flag in the main transaction execution logic to revert the transaction if necessary.

**Comment by hyal1ngrumblu** [comment]
The user is encountering an issue with their Paymaster.sol contract on the Testnet, where if any code inside the function postTransaction is reverted, the transaction itself does not get reverted but rather the entire postTransaction does not run. The user has not made any recent changes to their contract and is not interacting with any external libraries or contracts. The user has provided an example of the postTransaction function with a revert statement. The user is requesting assistance with forcing the transaction to fail if there is an error in postTransaction.
